### PR TITLE
attempt to avoid writing to disk and including redundant dependencies

### PR DESCRIPTION
**Pull request by cpsievert**
This is an attempt on an improvement to #81. There are two main benefits to an approach like this:

1. No pandoc necessary 🎉 
2. Resolves dependencies on the tag level, so redundant dependencies are never included (this can greatly reduce the size of the resulting notebook if it contains many widgets).

WARNING: I wrote this over a year ago and forgot to send this PR. I'm not actually sure if it currently works, but thought I would open to get a discussion going.

### CODE REVIEW

**Comment by takluyver** [comment]
Did you mean to check in the notebook?

File: R/repr_htmlwidget.r [hunk 1]
```diff
@@ -11,14 +110,14 @@ repr_text.htmlwidget <- function(obj, ...) 'HTML widgets cannot be represented i
...
-	htmlfile <- tempfile(fileext = '.html')
-	on.exit(unlink(htmlfile))
-	
-	htmlwidgets::saveWidget(obj, htmlfile)
-	
-	readChar(htmlfile, file.info(htmlfile)$size)
-}
+repr_html.htmlwidget <- embed_tags
+
+#' #' @name repr_*.shiny.tag
```
Comment (flying-sheep) [inline R/repr_htmlwidget.r]: This looks wrong. If you want to comment out things, please use `#`, as `#'` means “roxygen docs or directives”

File: Untitled1.ipynb [hunk 2]
```diff
@@ -0,0 +1,82 @@
+{
```
Comment (flying-sheep) [inline Untitled1.ipynb]: please remove the notebook 😄

**Comment by flying-sheep** [comment]
OK, cool! so one more unclear thing: why are there commented-out lines for `repr_html.shiny.tag`? will suporting those methods need changes to the code?

**Comment by cpsievert** [comment]
> so one more unclear thing: why are there commented-out lines for repr_html.shiny.tag? 

I forgot, to be honest, but I have a feeling I was waiting for #81, then lost motivation to contribute.  Anyway, it seems like `embed_tags()` could and should be used for `shiny.tag` and `shiny.tag.list`, so I think that part should be un-commented.

More importantly, now when I do something like:

```r
library(plotly)
plot_ly()
```

I just see:

```r
HTML widgets cannot be represented in plain text (need html)
```

Any idea why `repr_text.htmlwidget()` is getting called instead of `repr_html.htmlwidget()`?

**Comment by flying-sheep** [comment]
> Any idea why repr_text.htmlwidget() is getting called instead of repr_html.htmlwidget()?

both are called, but repr_html somehow failed or returned NULL apparently.

you could call it manually and step through in a debugger to see where it goes wrong.

> Anyway, it seems like embed_tags() could and should be used for shiny.tag and shiny.tag.list.

once the above bug is fixed, could you please check if they indeed do?

**Comment by cpsievert** [comment]
> repr_html somehow failed or returned NULL apparently

It returns `NULL` for some odd reason. By collapsing the character vector (see 0dfcdf7) it now returns the actual result (i.e., `cat(repr_html(plot_ly()), file = "index.html")` works), but for some reason the string is printed multiple times in the notebook ¯\_(ツ)_/¯

Also, after 403dc86145bd59592c821a33ad0fcd3495903459, I was able to confirm that `cat(repr_html(htmltools::tagList(plot_ly(), plot_ly())), file = "index.html")` works

**Comment by flying-sheep** [comment]
hm, something’s off: it only works once, then it return `""` :sweat_smile:

**Comment by flying-sheep** [comment]
Merged!

**Pull request closed without merging**